# Importing necessary libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

# Reading the data

In [2]:
cp = pd.read_csv(r'churn_prediction.csv')

In [3]:
cp.shape

(28382, 21)

In [4]:
cp.describe()

customer_id       vintage           age    dependents          city  \
count  28382.000000  28382.000000  28382.000000  25919.000000  27579.000000   
mean   15143.508667   2364.336446     48.208336      0.347236    796.109576   
std     8746.454456   1610.124506     17.807163      0.997661    432.872102   
min        1.000000    180.000000      1.000000      0.000000      0.000000   
25%     7557.250000   1121.000000     36.000000      0.000000    409.000000   
50%    15150.500000   2018.000000     46.000000      0.000000    834.000000   
75%    22706.750000   3176.000000     60.000000      0.000000   1096.000000   
max    30301.000000  12899.000000     90.000000     52.000000   1649.000000   

       customer_nw_category   branch_code  days_since_last_transaction  \
count          28382.000000  28382.000000                 25159.000000   
mean               2.225530    925.975019                    69.997814   
std                0.660443    937.799129                    86.341098   
min                1.000000      1.000000                     0.000000   
25%                2.000000    176.000000                    11.000000   
50%                2.000000    572.000000                    30.000000   
75%                3.000000   1440.000000                    95.000000   
max                3.000000   4782.000000                   365.000000   

       current_balance  previous_month_end_balance  \
count     2.838200e+04                2.838200e+04   
mean      7.380552e+03                7.495771e+03   
std       4.259871e+04                4.252935e+04   
min      -5.503960e+03               -3.149570e+03   
25%       1.784470e+03                1.906000e+03   
50%       3.281255e+03                3.379915e+03   
75%       6.635820e+03                6.656535e+03   
max       5.905904e+06                5.740439e+06   

       average_monthly_balance_prevQ  average_monthly_balance_prevQ2  \
count                   2.838200e+04                    2.838200e+04   
mean                    7.496780e+03                    7.124209e+03   
std                     4.172622e+04                    4.457581e+04   
min                     1.428690e+03                   -1.650610e+04   
25%                     2.180945e+03                    1.832507e+03   
50%                     3.542865e+03                    3.359600e+03   
75%                     6.666887e+03                    6.517960e+03   
max                     5.700290e+06                    5.010170e+06   

       current_month_credit  previous_month_credit  current_month_debit  \
count          2.838200e+04           2.838200e+04         2.838200e+04   
mean           3.433252e+03           3.261694e+03         3.658745e+03   
std            7.707145e+04           2.968889e+04         5.198542e+04   
min            1.000000e-02           1.000000e-02         1.000000e-02   
25%            3.100000e-01           3.300000e-01         4.100000e-01   
50%            6.100000e-01           6.300000e-01         9.193000e+01   
75%            7.072725e+02           7.492350e+02         1.360435e+03   
max            1.226985e+07           2.361808e+06         7.637857e+06   

       previous_month_debit  current_month_balance  previous_month_balance  \
count          2.838200e+04           2.838200e+04            2.838200e+04   
mean           3.339761e+03           7.451133e+03            7.495177e+03   
std            2.430111e+04           4.203394e+04            4.243198e+04   
min            1.000000e-02          -3.374180e+03           -5.171920e+03   
25%            4.100000e-01           1.996765e+03            2.074408e+03   
50%            1.099600e+02           3.447995e+03            3.465235e+03   
75%            1.357553e+03           6.667958e+03            6.654693e+03   
max            1.414168e+06           5.778185e+06            5.720144e+06   

              churn  
count  28382.000000  
mean       0.185329  
std        0.388571  
min        0.000000  
25% 

In [5]:
cp.head(5)

customer_id  vintage  age gender  dependents     occupation    city  \
0            1     3135   66   Male         0.0  self_employed   187.0   
1            2      310   35   Male         0.0  self_employed     NaN   
2            4     2356   31   Male         0.0       salaried   146.0   
3            5      478   90    NaN         NaN  self_employed  1020.0   
4            6     2531   42   Male         2.0  self_employed  1494.0   

   customer_nw_category  branch_code  days_since_last_transaction  ...  \
0                     2          755                        224.0  ...   
1                     2         3214                         60.0  ...   
2                     2           41                          NaN  ...   
3                     2          582                        147.0  ...   
4                     3          388                         58.0  ...   

   previous_month_end_balance  average_monthly_balance_prevQ  \
0                     1458.71                        1458.71   
1                     8704.66                        7799.26   
2                     5815.29                        4910.17   
3                     2291.91                        2084.54   
4                     1401.72                        1643.31   

   average_monthly_balance_prevQ2  current_month_credit  \
0                         1449.07                  0.20   
1                        12419.41                  0.56   
2                         2815.94                  0.61   
3                         1006.54                  0.47   
4                         1871.12                  0.33   

   previous_month_credit  current_month_debit  previous_month_debit  \
0                   0.20                 0.20                  0.20   
1                   0.56              5486.27                100.56   
2                   0.61              6046.73                259.23   
3                   0.47                 0.47               2143.33   
4                 714.61               588.62               1538.06   

   current_month_balance  previous_month_balance  churn  
0                1458.71                 1458.71      0  
1                6496.78                 8787.61      0  
2                5006.28                 5070.14      0  
3                2291.91                 1669.79      1  
4                1157.15                 1677.16      1  

[5 rows x 21 columns]

In [6]:
cp.isnull().sum()

customer_id                          0
vintage                              0
age                                  0
gender                             525
dependents                        2463
occupation                          80
city                               803
customer_nw_category                 0
branch_code                          0
days_since_last_transaction       3223
current_balance                      0
previous_month_end_balance           0
average_monthly_balance_prevQ        0
average_monthly_balance_prevQ2       0
current_month_credit                 0
previous_month_credit                0
current_month_debit                  0
previous_month_debit                 0
current_month_balance                0
previous_month_balance               0
churn                                0
dtype: int64

In [7]:
cp.dtypes

customer_id                         int64
vintage                             int64
age                                 int64
gender                             object
dependents                        float64
occupation                         object
city                              float64
customer_nw_category                int64
branch_code                         int64
days_since_last_transaction       float64
current_balance                   float64
previous_month_end_balance        float64
average_monthly_balance_prevQ     float64
average_monthly_balance_prevQ2    float64
current_month_credit              float64
previous_month_credit             float64
current_month_debit               float64
previous_month_debit              float64
current_month_balance             float64
previous_month_balance            float64
churn                               int64
dtype: object

In [8]:
cp['dependents'].value_counts()

0.0     21435
2.0      2150
1.0      1395
3.0       701
4.0       179
5.0        41
6.0         8
7.0         3
36.0        1
52.0        1
25.0        1
9.0         1
50.0        1
32.0        1
8.0         1
Name: dependents, dtype: int64

In [9]:
cp['occupation'].value_counts(normalize=True)

self_employed    0.617483
salaried         0.236874
student          0.072716
retired          0.071514
company          0.001413
Name: occupation, dtype: float64

In [10]:
cp['churn'].value_counts()

0    23122
1     5260
Name: churn, dtype: int64

In [11]:
cp.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 28382 entries, 0 to 28381
Data columns (total 21 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   customer_id                     28382 non-null  int64  
 1   vintage                         28382 non-null  int64  
 2   age                             28382 non-null  int64  
 3   gender                          27857 non-null  object 
 4   dependents                      25919 non-null  float64
 5   occupation                      28302 non-null  object 
 6   city                            27579 non-null  float64
 7   customer_nw_category            28382 non-null  int64  
 8   branch_code                     28382 non-null  int64  
 9   days_since_last_transaction     25159 non-null  float64
 10  current_balance                 28382 non-null  float64
 11  previous_month_end_balance      28382 non-null  float64
 12  average_monthly_balance_prevQ   

In [12]:
cp['days_since_last_transaction'].value_counts().head(10)

0.0     1672
3.0      831
14.0     654
4.0      632
6.0      596
5.0      576
7.0      561
11.0     486
13.0     478
9.0      476
Name: days_since_last_transaction, dtype: int64

# Data Preprocessing

In [15]:
cp.drop("customer_id", axis=1, inplace=True)

In [16]:
cp.gender.replace(["Male","Female"], [1,0], inplace=True)

In [17]:
cp['gender'] = cp['gender'].fillna(cp['gender'].mode()[0])

In [18]:
cp['dependents'] = cp['dependents'].fillna(cp['dependents'].mode()[0])

In [19]:
cp

vintage  age  gender  dependents     occupation    city  \
0         3135   66     1.0         0.0  self_employed   187.0   
1          310   35     1.0         0.0  self_employed     NaN   
2         2356   31     1.0         0.0       salaried   146.0   
3          478   90     1.0         0.0  self_employed  1020.0   
4         2531   42     1.0         2.0  self_employed  1494.0   
...        ...  ...     ...         ...            ...     ...   
28377     1845   10     0.0         0.0        student  1020.0   
28378     4919   34     0.0         0.0  self_employed  1046.0   
28379      297   47     1.0         0.0       salaried  1096.0   
28380     2585   50     1.0         3.0  self_employed  1219.0   
28381     2349   18     1.0         0.0        student  1232.0   

       customer_nw_category  branch_code  days_since_last_transaction  \
0                         2          755                        224.0   
1                         2         3214                         60.0   
2                         2           41                          NaN   
3                         2          582                        147.0   
4                         3          388                         58.0   
...                     ...          ...                          ...   
28377                     2         1207                         70.0   
28378                     2          223                         14.0   
28379                     2          588                          0.0   
28380                     3          274                          NaN   
28381                     2          474                         59.0   

       current_balance  previous_month_end_balance  \
0              1458.71                     1458.71   
1              5390.37                     8704.66   
2              3913.16                     5815.29   
3              2291.91                     2291.91   
4               927.72                     1401.72   
...                ...                         ...   
28377          1076.43                     1076.43   
28378          3844.10                     4069.21   
28379         65511.97                    61017.55   
28380          1625.55                     1625.55   
28381          2107.05                     2821.34   

       average_monthly_balance_prevQ  average_monthly_balance_prevQ2  \
0                            1458.71                         1449.07   
1                            7799.26                        12419.41   
2                            4910.17                         2815.94   
3                            2084.54                         1006.54   
4                            1643.31                         1871.12   
...                              ...                             ...   
28377                        2282.19                         2787.70   
28378                        3668.83                         3865.55   
28379                       53444.81                        21925.81   
28380                        1683.20                         1857.42   
28381                        3213.44                         4447.45   

       current_month_credit  previous_month_credit  current_month_debit  \
0                      0.20                   0.20                 0.20   
1                      0.56                   0.56              5486.27   
2                      0.61                   0.61              6046.73   
3                      0.47                   0.47                 0.47   
4                      0.33                 714.61               588.62   
...                     ...                    ...                  ...   
28377                  0.30                   0.30                 0.30   
28378                  1.71                   2.29               901.00   
28379               4666.84                3883.06               168.23   
28380                  0.20                   0.20                 0.20  

In [20]:
cp['occupation'] = cp['occupation'].fillna('self_employed')

In [21]:
cp.isnull().sum()

vintage                              0
age                                  0
gender                               0
dependents                           0
occupation                           0
city                               803
customer_nw_category                 0
branch_code                          0
days_since_last_transaction       3223
current_balance                      0
previous_month_end_balance           0
average_monthly_balance_prevQ        0
average_monthly_balance_prevQ2       0
current_month_credit                 0
previous_month_credit                0
current_month_debit                  0
previous_month_debit                 0
current_month_balance                0
previous_month_balance               0
churn                                0
dtype: int64

In [22]:
cp['days_since_last_transaction'] = cp['days_since_last_transaction'].fillna(cp['days_since_last_transaction'].mean())

In [23]:
cp['city'] = cp['city'].fillna(cp['city'].mean())

In [25]:
from sklearn.preprocessing import LabelEncoder
lb = LabelEncoder()
cp['occupation'] = lb.fit_transform(cp['occupation'])

In [26]:
cp.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 28382 entries, 0 to 28381
Data columns (total 20 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   vintage                         28382 non-null  int64  
 1   age                             28382 non-null  int64  
 2   gender                          28382 non-null  float64
 3   dependents                      28382 non-null  float64
 4   occupation                      28382 non-null  int32  
 5   city                            28382 non-null  float64
 6   customer_nw_category            28382 non-null  int64  
 7   branch_code                     28382 non-null  int64  
 8   days_since_last_transaction     28382 non-null  float64
 9   current_balance                 28382 non-null  float64
 10  previous_month_end_balance      28382 non-null  float64
 11  average_monthly_balance_prevQ   28382 non-null  float64
 12  average_monthly_balance_prevQ2  

# Model Building

In [27]:
X=cp.drop("churn",axis=1)
y=cp.churn

In [28]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=32)
print(X_train.shape, y_train.shape, X_test.shape, y_test.shape)

(19867, 19) (19867,) (8515, 19) (8515,)


In [29]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score

In [30]:
rf = RandomForestClassifier(n_estimators=100, max_depth=20, random_state=40)
rf.fit(X_train, y_train)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=20, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=100,
                       n_jobs=None, oob_score=False, random_state=40, verbose=0,
                       warm_start=False)

# Testing the model

In [31]:
score = rf.score(X_train, y_train)
score2 = rf.score(X_test, y_test)
score, score2

(0.9706045200583883, 0.8650616559013505)

In [32]:
y_pred = rf.predict(X_test)

In [33]:
print(confusion_matrix(y_test, y_pred))

[[6639  300]
 [ 849  727]]


In [34]:
print(accuracy_score(y_test, y_pred))

0.8650616559013505


In [35]:
fe = pd.DataFrame({'features':list(X_train.columns), 'importance':rf.feature_importances_}).sort_values('importance', ascending=False)

In [62]:
fe.head(10)

features  importance
9                  current_balance    0.205777
17           current_month_balance    0.087602
15             current_month_debit    0.083829
16            previous_month_debit    0.065349
12  average_monthly_balance_prevQ2    0.061457
10      previous_month_end_balance    0.059958
11   average_monthly_balance_prevQ    0.059761
18          previous_month_balance    0.057168
0                          vintage    0.046788
13            current_month_credit    0.044539

# Decision tree Classifier

In [48]:
from sklearn.tree import DecisionTreeClassifier
dt_clf = DecisionTreeClassifier(criterion='entropy', min_samples_split=200)

In [49]:
dt_clf.fit(X_train, y_train)

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='entropy',
                       max_depth=None, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=200,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=None, splitter='best')

In [50]:
socre = accuracy_score(y_train, dt_clf.predict(X_train)) #training accuracy
score1 = accuracy_score(y_test, dt_clf.predict(X_test)) #testing accuracy

In [51]:
score, score1

(0.9706045200583883, 0.851556077510276)

# KNN

In [52]:
from sklearn.preprocessing import StandardScaler

sc = StandardScaler().fit(X_train)
X_train_std = sc.transform(X_train)
X_test_std = sc.transform(X_test)

In [58]:
from sklearn.neighbors import KNeighborsClassifier

knn = KNeighborsClassifier(n_neighbors=17)
knn.fit(X_train_std,y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
                     metric_params=None, n_jobs=None, n_neighbors=17, p=2,
                     weights='uniform')

In [59]:
#prediction of the KNN model
predict_knn = knn.predict(X_test_std)
accuracy_knn = accuracy_score(y_test,predict_knn)*100
print("Accuracy of the KNN classifier is:",accuracy_knn) #testing accuracy

Accuracy of the KNN classifier is: 81.5619495008808


In [61]:
accuracy_score(y_train, knn.predict(X_train_std))*100 #training accuracy

81.6177580913072

###  As we can see from the above 3 models that KNN algorithm gives us the best fit model with training accuracy of 81.6% and testing accuracy of 81.5%